### PR TITLE
Implement combineMetrics for StakePools

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#711 

# Overview

- [x] I have implemented `combineMetrics`, combining stake-distribution with pool-productions.

# Comments

### CODE REVIEW

File: lib/core/test/unit/Cardano/Pool/MetricsSpec.hs [hunk 1]
```diff
@@ -47,6 +49,24 @@ spec = do
...
+                property $ \stakeDistr poolProd ->
+                    let
+                        aPoolWithoutStakeProduced =
+                            not . Map.null $ Map.difference poolProd stakeDistr
+                    in classify aPoolWithoutStakeProduced
+                        "A pool without stake produced" $
+                        combineMetrics stakeDistr poolProd
+                        `shouldSatisfy` \case
+                            Left _ -> aPoolWithoutStakeProduced
+                            Right _x -> True
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Pool/MetricsSpec.hs]: These two properties, I think, are the _particularly_ interesting cases… although maybe a bit thin…

File: lib/core/src/Cardano/Pool/Metrics.hs [hunk 2]
```diff
@@ -40,6 +49,52 @@ activityForEpoch epoch s =
...
+    :: Map PoolId (Quantity "lovelace" Word64)
+    -> Map PoolId Int
+    -> Either
+        ErrMetricsInconsistency
+        (Map PoolId (Quantity "lovelace" Word64, Int))
+combineMetrics =
+    mergeA
+        stakeButNoActivity
+        activityButNoStake
+        (zipWithMatched (\_k stake productions -> (stake, productions)))
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metrics.hs]: I like how the first two cases are made very clear with the intermediary names. I'd have carried on for the 3rd case too:

```hs
mergeA stakeButNoActivity activityButNoStake stakeAndActivity
```

File: lib/core/src/Cardano/Pool/Metrics.hs [hunk 3]
```diff
@@ -40,6 +49,52 @@ activityForEpoch epoch s =
...
+  where
+    -- What to do when we only find a pool in one of the two. data-sources:
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metrics.hs]: That comment seems slightly off :thinking: ?

File: lib/core/test/unit/Cardano/Pool/MetricsSpec.hs [hunk 4]
```diff
@@ -47,6 +49,24 @@ spec = do
                 let count = Map.foldl (\r l -> r + (length l)) 0 . activity
                 count s' === (count s + 1)
 
+        describe "combineMetrics" $ do
+            it "pools with no entry for productions are included" $
+                property $ \stakeDistr -> do
+                    combineMetrics stakeDistr Map.empty
+                    `shouldBe`
+                    (Right $ Map.map (, 0) stakeDistr)
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Pool/MetricsSpec.hs]: :+1:

File: lib/core/test/unit/Cardano/Pool/MetricsSpec.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Pool/MetricsSpec.hs]: I think it gives a nice coverage and a good sense of what we expect.
What about some coverage check for the classification to make sure that we are actually testing both cases?

**Review by KtorZ — APPROVED** [review]

**Pull request merged**